### ISSUE TITLE
.NET6 rc - MSB3030: Azure Pipeline Publish fails for Blazor Server Project that references nested razor component libraries

### ISSUE THREAD

**Issue by ameyasubhedar** [dotnet/aspnetcore#36639]
Warly Evaluation: .NET 6 rc1

## Environemnt
.NET SDK 6.0.100-rc.1
Full version
6.0.100-rc.1.21458.32

Visual Studio 2022 Preview 4 (Latest Preview)

Azure Devops Pipeline

ASP.NET
Blazor Server
Blazor WebAssembly
Razor Component Libraries

### Describe the bug
Azure Pipeline Build and Publish fails to publish the Blazor Server projects that reference a 2 level razor component projects.

### Sample Blazor structure with nested razor component library
```
TestC.Web.Server.csproj  (Blazor Server Project)
|
|--> TestC.Data.Components.csproj  (Project Reference - Razor Component Library) 
     |
     |--> TestC.Shared.Components.csproj  (Project Referece - Razor Component Library)
          |
          |-- wwwroot (css, img, js)
|
| -- Server Files
| -- wwwroot (server)
```

**_NOTE_**  All the UX files css, img and js are part of TestC.Shared.Components razor component library. But the error seems to suggest that the TestC.Web.Server publish failed because the same files (e.g.  css\centeredlayout.min.css) were not found by project TestC.Data.Components. All the files that publish failed actually belongs to TestC.Shared.Components.csproj but somehow the publish command is trying to find those in TestC.Data.Components.csproj.

**_The same project published successfully on local machine._**

**_This was not an issue with preview6_**

### Error 
`
C :\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\centeredlayout.min.css" because it was not found. `

### Local Machine (Success)
```
** Visual Studio 2022 Developer PowerShell v17.0.0-pre.4.0
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
PS D:\this\PI\Services> cd src
PS D:\this\PI\Services\src> cd TestC.Web.Server
PS D:\this\PI\Services\src\TestC.Web.Server> dotnet publish TestC.Web.Server.csproj --configuration Release
Microsoft (R) Build Engine version 17.0.0-preview-21458-01+2c5510013 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored D:\this\PI\Services\src\TestC.Shared\TestC.Shared.csproj (in 913 ms).
  Restored D:\this\PI\Services\src\TestC.Protocols.Defaults\TestC.Protocols.Defaults.csproj (in 913 ms).
  Restored D:\this\PI\Services\src\TestC.Shared.Components\TestC.Shared.Components.csproj (in 912 ms).
  Restored D:\this\PI\Services\src\TestC.Protocols\TestC.Protocols.csproj (in 913 ms).
  Restored D:\this\PI\Services\src\TestC.Data.Shared\TestC.Data.Shared.csproj (in 913 ms).
  Restored D:\this\PI\Services\src\TestC.Data.Components\TestC.Data.Components.csproj (in 1.3 sec).
  Restored D:\this\PI\Services\src\TestC.Data\TestC.Data.csproj (in 1.3 sec).
  Restored D:\this\PI\Services\src\TestC.Imlx\TestC.Imlx.csproj (in 1.49 sec).
  Restored D:\this\PI\Services\src\TestC.Imlx.DotNet\TestC.Imlx.DotNet.csproj (in 1.49 sec).
  1 of 10 projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  TestC.Shared -> D:\this\PI\Services\src\TestC.Shared\bin\Release\netstandard2.1\TestC.Shared.dll
  TestC.Data.Shared -> D:\this\PI\Services\src\TestC.Data.Shared\bin\Release\netstandard2.1\TestC.Data.Shared.dll

  Bundler: Begin processing bundleconfig.json
  TestC.Protocols -> D:\this\PI\Services\src\TestC.Protocols\bin\Release\netstandard2.1\TestC.Protocols.dll
  Bundler: Done processing bundleconfig.json
  TestC.Shared.Components -> D:\this\PI\Services\src\TestC.Shared.Components\bin\Release\net6.0\TestC.Shared.Components.dll
  TestC.Data -> D:\this\PI\Services\src\TestC.Data\bin\Release\netstandard2.1\TestC.Data.dll
  TestC.Data.Components -> D:\this\PI\Services\src\TestC.Data.Components\bin\Release\net6.0\TestC.Data.Components.dll
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  TestC.Imlx -> D:\this\PI\Services\src\TestC.Imlx\bin\Release\net6.0\TestC.Imlx.dll
  TestC.Imlx.DotNet -> D:\this\PI\Services\src\TestC.Imlx.DotNet\bin\Release\net6.0\TestC.Imlx.DotNet.dll
  TestC.Protocols.Defaults -> D:\this\PI\Services\src\TestC.Protocols.Defaults\bin\Release\netstandard2.1\TestC.Protocols.Defaults.dll

  Bundler: Begin processing bundleconfig.json
  Bundler: Done processing bundleconfig.json
  TestC.Web.Server -> D:\this\PI\Services\src\TestC.Web.Server\bin\Release\net6.0\TestC.Web.Server.dll
  TestC.Web.Server -> D:\this\PI\Services\src\TestC.Web.Server\bin\Release\net6.0\publish\
PS D:\this\PI\Services\src\TestC.Web.Server>
```

### Azure Pipeline (Failure)
```
2021-09-17T01:48:29.8453536Z ##[section]Starting: Publish
2021-09-17T01:48:29.8708619Z ==============================================================================
2021-09-17T01:48:29.8708925Z Task         : .NET Core
2021-09-17T01:48:29.8709295Z Description  : Build, test, package, or publish a dotnet application, or run a custom dotnet command
2021-09-17T01:48:29.8709563Z Version      : 2.187.0
2021-09-17T01:48:29.8709731Z Author       : Microsoft Corporation
2021-09-17T01:48:29.8710281Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/build/dotnet-core-cli
2021-09-17T01:48:29.8710585Z ==============================================================================
2021-09-17T01:48:30.3648118Z [command]C:\Windows\system32\chcp.com 65001
2021-09-17T01:48:30.3845900Z Active code page: 65001
2021-09-17T01:48:30.3940560Z Info: .NET Core SDK/runtime 2.2 and 3.0 are now End of Life(EOL) and have been removed from all hosted agents. If you're using these SDK/runtimes on hosted agents, kindly upgrade to newer versions which are not EOL, or else use UseDotNet task to install the required version.
2021-09-17T01:48:30.8208435Z [command]C:\hostedtoolcache\windows\dotnet\dotnet.exe publish D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj --configuration Release --output D:\a\9\a\TestC.Web.Server
2021-09-17T01:48:31.0272080Z Microsoft (R) Build Engine version 17.0.0-preview-21458-01+2c5510013 for .NET
2021-09-17T01:48:31.0282940Z Copyright (C) Microsoft Corporation. All rights reserved.
2021-09-17T01:48:31.0283319Z 
2021-09-17T01:48:31.6915163Z   Determining projects to restore...
2021-09-17T01:48:32.0441538Z   All projects are up-to-date for restore.
2021-09-17T01:48:32.2606407Z   You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
2021-09-17T01:48:32.6424272Z   
2021-09-17T01:48:32.6424938Z   Bundler: Begin processing bundleconfig.json
2021-09-17T01:48:32.7921383Z   Bundler: Done processing bundleconfig.json
2021-09-17T01:48:32.9664399Z   TestC.Web.Server -> D:\a\9\s\src\TestC.Web.Server\bin\Release\net6.0\TestC.Web.Server.dll
2021-09-17T01:48:33.3023902Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\centeredlayout.min.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3028836Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\centeredlayout.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3044001Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\elements.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3045937Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\components.min.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3049827Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\style.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3051782Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\style.min.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3053379Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\css\elements.min.css" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3055010Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\img\logo\pidot-64.png" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3057695Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\img\logo\pidot-128.png" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3061978Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\js\style.js" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3064053Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\img\logo\pidot.png" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3066187Z C:\hostedtoolcache\windows\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.StaticWebAssets.targets(612,5): error MSB3030: Could not copy the file "C:\Users\VssAdministrator\.nuget\packages\TestC.data.components\1.2.4-preview.21091703\staticwebassets\js\style.min.js" because it was not found. [D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj]
2021-09-17T01:48:33.3458092Z ##[error]Error: The process 'C:\hostedtoolcache\windows\dotnet\dotnet.exe' failed with exit code 1
2021-09-17T01:48:33.3460413Z ##[warning].NET 5 has some compatibility issues with older Nuget versions(<=5.7), so if you are using an older Nuget version(and not dotnet cli) to restore, then the dotnet cli commands (e.g. dotnet build) which rely on such restored packages might fail. To mitigate such error, you can either: (1) - Use dotnet cli to restore, (2) - Use Nuget version 5.8 to restore, (3) - Use global.json using an older sdk version(<=3) to build
2021-09-17T01:48:33.3464489Z Info: Azure Pipelines hosted agents have been updated and now contain .Net 5.x SDK/Runtime along with the older .Net Core version which are currently lts. Unless you have locked down a SDK version for your project(s), 5.x SDK might be picked up which might have breaking behavior as compared to previous versions. You can learn more about the breaking changes here: https://docs.microsoft.com/en-us/dotnet/core/tools/ and https://docs.microsoft.com/en-us/dotnet/core/compatibility/ . To learn about more such changes and troubleshoot, refer here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/build/dotnet-core-cli?view=azure-devops#troubleshooting
2021-09-17T01:48:33.3467270Z ##[error]Dotnet command failed with non-zero exit code on the following projects : D:\a\9\s\src\TestC.Web.Server\TestC.Web.Server.csproj
2021-09-17T01:48:33.3571416Z ##[section]Finishing: Publish
```

**Comment by javiercn** [comment]
@ameyasubhedar thanks for contacting us.

Your description below suggest that the project is using project references, however in your AzDo pipeline it looks like those references are being consumed as package references instead of project references.

The only way for that to happen is that you have a PackageReference (instead of a project reference) and that your pipeline is building the project into a package and having the build consume it. 

The error seems to indicate that the files are being consumed as if they came from a package and that they are not present on the packages folder.

In order for us to make progress on this issue we would ideally need a minimal repro project and detailed steps to reproduce the issue or at least a binary log file generated for the failing project that we can inspect to get a sense of what's going on in your build process.

```
TestC.Web.Server.csproj  (Blazor Server Project)
|
|--> TestC.Data.Components.csproj  (Project Reference - Razor Component Library) 
     |
     |--> TestC.Shared.Components.csproj  (Project Referece - Razor Component Library)
          |
          |-- wwwroot (css, img, js)
|
| -- Server Files
| -- wwwroot (server)
```

**Comment by ameyasubhedar** [comment]
@javiercn, Yes on local machine they are the project reference. But in the Azure DevOps the pipeline is builing the class libraries into packages and then the server and top packages consumes it in the same build as Nuget Package references.

**Local Machine**

```
TestC.Web.Server.csproj  (Blazor Server Project)
|
|--> TestC.Data.Components.csproj  (Project Reference - Razor Component Library) 
     |
     |--> TestC.Shared.Components.csproj  (Project Referece - Razor Component Library)
          |
          |-- wwwroot (css, img, js)
|
| -- Server Files
| -- wwwroot (server)
```

**Azure Pipeline**

TestC.Shared.Components.csproj  -> NugetPackage -> Azure Artifact
TestC.Data.Components.csproj -> Nuget Package -> Azure Artifact

```
TestC.Web.Server.csproj  (Blazor Server Project)
|
|--> TestC.Data.Components.csproj  (Nuget Package Reference - Razor Component Library) 
     |
     |--> TestC.Shared.Components.csproj  (Nuget Package Referece - Razor Component Library)
          |
          |-- wwwroot (css, img, js)
|
| -- Server Files
| -- wwwroot (server)
```

I acheive by defining the Azure Pipeline Variables $(ENVPREVIEW) and then using them in the TestC.Web.Server and TestC.Data.Components  csproj file.

**_The same setup worked in .NET5_**

```
<Choose>
        <!--Preview Pipeline-->
        <When Condition="'$(ENVPREVIEW)'=='Preview'">
            <ItemGroup>
                <PackageReference Include="TestC.Data.Components" Version="$(ENVPREVIEWVER)" />
            </ItemGroup>
        </When>
     
        <!--Undefined/Local-->
        <Otherwise>
            <ItemGroup>
                <ProjectReference Include="..\TestC.Data.Components\TestC.Data.Components.csproj" />            
            </ItemGroup>
        </Otherwise>
    </Choose>
```

**Comment by javiercn** [comment]
@ameyasubhedar thanks for the additional details. 

This seems like a packing issue.

For us to make progress we would need the following things:
1. For you to grab the package from your CI pipeline and to check the contents to see if the files are present on the package.
2. If you determine that the files are not present, then produce a minimal repro project (as a public github repo) that demonstrates the packing issue, so that we can get an exact understanding of how you are creating the package, identify any potential issue and provide solutions.

**Comment by Valks** [comment]
I think I've been running into the same (root) issue. I haven't been able to reproduce the problem reliably locally. After updating my local copy of dotnet (WSL) from preview.7 to rc.1 it stopped showing the behaviour.

When building a Razor Class Library the staticwebassets folder isn't been populated in the Nuget package if the file isn't in the ```wwwroot``` folder. When I was able to locally reproduce the issue Windows (Visual Studio or cmd line) would build correctly but WSL would exhibit the same behaviour I'm seeing building via CI (Azure DevOps).

An interesting behaviour I noticed (locally and Azure DevOps) is that it still populates the build/Microsoft.AspNetCore.StaticWebAssets.props file even though the files haven't been added. Also, if the files are in the ```wwwroot``` folder they are added correctly whilst those outside are listed but not added (but still in the .props file).

I have built the repo below as a sample to try and reproduce the issue but it's slightly different again from the main project I'm having issues with. For some reason, the GitHub actions version is populating a ```content``` and ```contentFiles``` folder.

https://github.com/Valks/RCLStaticResourceTest

Also attached is the same project built locally on my PC, as can be seen inside the NuGet package the staticwebapps folder is populated correctly.

[RCLStaticResourceTest.1.0.0.zip](https://github.com/dotnet/aspnetcore/files/7194024/RCLStaticResourceTest.1.0.0.zip)

The pictures below are what I'm experiencing after building my main project via CI/CD.

![image](https://user-images.githubusercontent.com/3953780/133967120-fb71fad9-451e-4f08-96cd-a0b091e54e7d.png)
![image](https://user-images.githubusercontent.com/3953780/133967133-b0bf9044-362e-49dc-8416-5fd6c5949180.png)
